### ISSUE TITLE
sync-ebooks update question

### ISSUE THREAD

**Issue by vr8hub** [standardebooks/web#78]
The sync-ebooks script has a two-step process: first, update any existing directories, then clone any new repositories.

In the update portion, it updates each directory by doing a `git fetch`. As I'm sure you know, but I didn't (I've learned enough `git` to use it for SE, but I've never seen or needed a `fetch`), fetch gets the changes but doesn't update the working directory; you have to do a subsequent `git merge` to do that.

I discovered this because I ran a sync tonight and didn't get any new changes. Or at least it looked that way to me. After doing some investigation, I found that I had the changes, I just had to do a `git merge` in every directory to see them. Which seems… inefficient.

My first question is: is that for purposes of bare repositories, i.e. should only a fetch be done when updating a server?

My next question is: for the general use case of keeping a set of WD clones, wouldn't we want the script to do both the `fetch` and the `merge` (or a `pull`)? Is there any reason _not_ to do that? (From what I've read, using fetch/merge is "safer" than doing a "git pull", although if we're going to do both automatically I don't know that it makes a difference.)

**Comment by acabal** [comment]
Well, `git merge` may introduce merge conflicts. That's generally why the two operations are separate. `git pull` is (basically) a kind of alias for `git fetch` directly followed by `git merge`.

With only `git fetch`, we can inspect the differences between the repos before attempting a merge. Though, I assume you would want to merge upstream changes anyway, if you're syncing the repos. The only question is what do we do in the case of a merge conflict. The script would have to alert the user that a merge conflict exists in xyz repos after syncing.

Also, we want to avoid merge commits. I don't know what the formula is for `git fetch` + `git merge`, but `git pull --rebase` rebases the repo on top of upstream changes, instead of creating merge commits. That's what we want to accomplish, regardless of what we do.

**Comment by vr8hub** [comment]
Right. I guess the question is what do we want the _script_ to do. IOW, if someone is using the script to sync everything, isn't it because they want everything synced? Why would there be merge conflicts? If they're actively working on one of the ~400 repositories, then maybe they wouldn't be using the script to sync everything until they were done? (I'm asking; I'm not doing any with these but having them there for testing; frankly, if I _did_ make changes, I want them to go away if I resync. But I don't know how anyone else uses a full synced WD copy.)

Alternatively, if that is commonly done (making changes to the bulk-synced versions of the repositories while still wanting to sync everything and keep the changes), maybe we have the script check for the presence of a (empty) file, either in the repository or at the top level of the sync directory, that tells the script to skip doing a merge (or a fetch?)

Bottom line, it's easy enough for me to just run a git merge on every directory every time I sync. I'm just trying to figure out if we can make things easier on everyone else (and me) by making a change to the script that will do the right thing and not do the wrong thing. Maybe a `--pull` flag that says use `git pull --rebase` instead of `git fetch`? The default would do what it does now, `fetch` only, using the flag would do a `pull --rebase` instead.

**Comment by acabal** [comment]
Merge conflicts can happen (they have to me) and it's an important error case we have to consider. I think it would be helpful if the script was updated to do a `git pull --rebase` automatically; but, it should silence git's pull output *unless* either --verbose is invoked, or there is a merge conflict, in which case the script should alert the user.

**Comment by vr8hub** [comment]
All right, I'll need to create some merge conflicts (with my repositories, not SE's :)) and see how best to handle them. I'll work on it.

**Comment by acabal** [comment]
Closed due to inactivity
